crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -566,6 +616,155 @@ impl Object {
+                    }
+
+                    let left = snapshot::objects_snapshot
+                        .select(snapshot::checkpoint_sequence_number)
+                        .order(snapshot::checkpoint_sequence_number.desc())
+                        .limit(1);
+
+                    historical_query = historical_query.filter(
+                        history::checkpoint_sequence_number.ge(coalesce(left.single_value(), 0)),
+                    );
+
+                    // if some checkpoint_sequence_number is provided, use it as the upper bound.
+                    // This is because an object can be extant at some target, but be indexed with
+                    // at a checkpoint_sequence_number <= target.
+                    if let Some(checkpoint_sequence_number) = checkpoint_sequence_number {
+                        // A validation check is not needed as we coalesce both sides to 0 for null
+                        // values, and if left > right, the db will return 0 rows. This also saves
+                        // us a roundtrip.
+                        historical_query = historical_query.filter(
+                            history::checkpoint_sequence_number.le(checkpoint_sequence_number),
+                        );
+
+                        snapshot_query = snapshot_query.filter(
+                            snapshot::checkpoint_sequence_number.le(checkpoint_sequence_number),
+                        );
+                    } else {
+                        let right = checkpoints::checkpoints
+                            .select(checkpoints::sequence_number)
+                            .order(checkpoints::sequence_number.desc())
+                            .limit(1);
+
+                        historical_query = historical_query.filter(
+                            history::checkpoint_sequence_number
+                                .le(coalesce(right.single_value(), 0)),
+                        );
+                    }
+
+                    snapshot_query.union(historical_query)
+                })
+                .optional()
+            })
+            .await?;
+
+        // For the moment, if the object existed at some point, it will have eventually be written to objects_snapshot.
+        // Therefore, if both results are None, the object has never existed.
+        let Some(stored_objs) = results else {
+            return Ok(None);
+        };
+
+        match stored_objs
+            .iter()
+            // First, filter the objects based on an exact match on version and
+            // checkpoint_sequence_number if provided
+            .filter(|stored_obj| {
+                version.map_or(true, |version| stored_obj.object_version == version)
+                    && checkpoint_sequence_number.map_or(true, |checkpoint| {
+                        stored_obj.checkpoint_sequence_number <= checkpoint
+                    })
```

Comment:
Why is this filtering necessary? I would have thought the filters you applied to the DB query would have sorted this out?